### ISSUE TITLE
Cascading Shadow Mapping with ProjectedTextures, requiring a missing element.

### ISSUE THREAD

**Issue by xoticdev** [Facepunch/garrysmod-issues#4455]
Hello, this is one of my favorite games thank you for your time. I noticed a major issue with certain maps where Shadow detail wasn't up to a dynamic standard. We have Garry's Mod 64bit now, but shadow casting always needed to be able to change sun angles during runtime. This is a costly render process, and CSM is not recommended for small maps, but for larger environments. . Yes, we have BSP baked shadows, and with have "basic shadows" which does a decent job. However, Cascading Shadow Mapping would provide Garry's Mod ultimate control over shadow quality.

A CSM method is a shadow shading technique used for larger landscapes and environments to provide dynamic shadows
![image](https://user-images.githubusercontent.com/63826034/79520575-333cf680-8025-11ea-9e6c-b8ad110ab8de.png)

The photo above is using Projected Textures to draw the shadows on the top left, however looking at the far left hand side you'll notice that the CSM is done incorrectly. The reason for this is as follows: If I set the CSM matrix to follow the PITCH of the camera, the next photo shows a PROPER CSM:
![image](https://user-images.githubusercontent.com/63826034/79520880-cbd37680-8025-11ea-86bd-427f824e5ef2.png)

Now you will notice in the second picture, the entire scene is completely covered by the Orthographic Matrices drawing as they should in a CSM within any game engine (DirectX, or Vulkan).

I set the Cascades into different colors so you can take a look.

The problem is that we're having clipping problems. You'll notice that the shadows level of detail (Smaller == higher quality) are overlapping the other larger layers. To complete Cascading Shadow maps, within the GLSL shaders (DirectX's shaders) one would use what us Garry's Mod Lua Coders call a [render.DepthRange](https://wiki.facepunch.com/gmod/render.DepthRange). 

If Garry's Mod had given me access, I'd add another function to [ProjectedTexture](https://wiki.facepunch.com/gmod/Global.ProjectedTexture) called DepthRange.

Doing so, would allow for me to cull out the layers based on depth. It's a hard thing to explain but I'll demonstrate what I mean:

![image](https://user-images.githubusercontent.com/63826034/79521635-a7789980-8027-11ea-981e-670b0c9f8071.png)

The picture above shows a great example, depending on the thin depth test, one can eliminate the extra chunks of useless ProjectedTexture clipping. This would end up making the shadows look seamless as appose to the first picture shown, it covers the entire camera's view frustum. The problem is that the layering of projected textures makes the shadows look fake. Now I know what you're thinking ProjectedTextures can be swapped for a circle radiant and would make shadows look better. Now it is a viable option to cope without this feature, shadows would look 100% realistic with the suggestion I have provided vs using a hacky method to render the shadows. Also, this would allow the maps to have **moving shadows** via time lapses.

Another photo of the overlapping cascades:
![image](https://user-images.githubusercontent.com/63826034/79522073-c0ce1580-8028-11ea-9ee2-e59d6cff3b92.png)

Without a depth test, the best option is to avoid using the pitch system which defeats majority of CSMs. When on the ground, it looks okay, but still noticeable.

![image](https://user-images.githubusercontent.com/63826034/79522204-14406380-8029-11ea-9f81-b51947121ce7.png)
Although the above picture looks decent, you'll notice the seams overlap causing a distortion of light. This would be solved since only 1 ProjectedTexture would display on that depth amount.

I want to be able to control the depth range of ProjectedTexturess so that the CSM will look blended correctly, making the shadows look higher detailed. 

![image](https://user-images.githubusercontent.com/63826034/79522494-e27bcc80-8029-11ea-83fd-907e06bc54c6.png)
The last photo is CSM in a different game, except this time there is a depth clamp, causing seamless shadows. This is ideal for shadow casting, but it's a good reference of how the shadows would look after the depth clamping.

Lastly I wanna offer you an example code of how it may or may not work in theory:
```
local self = LocalPlayer()
local lamp = ProjectedTexture()
	self.lamp = lamp
	lamp:SetTexture( "effects/flashlight/hard" )
	lamp:SetFarZ( 500 )
	//lamp:SetNoDraw();      Disables drawing light normally
	lamp:SetPos( self:GetPos() )
	lamp:SetAngles( self:GetAngles() )
	lamp:Update()

local lamp2 = ProjectedTexture()
	self.lamp = lamp
	lamp2:SetTexture( "effects/flashlight/hard" )
	lamp2:SetFarZ( 500 )
	//lamp2:SetNoDraw();      Disables drawing light normally
	lamp2:SetPos( self:GetPos() )
	lamp2:SetAngles( self:GetAngles() )
	lamp2:Update()


hook.Add("PreDrawOpaqueRenderables", "ThisIsATest", function()
	lamp:SetPos( self:GetPos() + Vector( 0,0,100) )
	lamp:SetAngles( self:GetAngles() )
	lamp:Update()

	lamp2:SetPos( self:GetPos() + Vector(0,0,100) - self:GetAngles():Forward() * -100 )
	lamp2:SetAngles( self:GetAngles() )
	lamp2:Update()

	// ProjectedTexture has a depth range so that Layering for Cascading shadow maps is possible.
	// I.E. Selecting a specific lamp, and clamping it's drawing according to DEPTH...
	
	render.DepthRange(0.0, 0.5); // THIS is important, with this, the layering can be possible.
		lamp:DrawModel()
	render.DepthRange(0.5, 1.0); // THIS is important, with this, the layering can be possible.
		lamp2:DrawModel()
end )
```

If you can get this added in, I'll 100% share my CSM code so that way people can do stuff like this:

https://youtu.be/gwgjKxcPZ1o?t=27

Thank you for your time, I will actively strive to get Garry's Mod higher quality shadows, if you allow me to proceed!!!! I appreciate the team that makes this game possible though out the years!!

**Comment by xoticdev** [comment]
I apologize for posting it here, I couldn't find the request place. But if you want I can remake the topic there, but this is important to me. In all seriousness. I need this to work.

**Comment by xoticdev** [comment]
I did some digging looking into ProjectedTexture content, and I found a [video of someone else trying to do what I'm doing](https://www.youtube.com/watch?v=4RtnCT6GAjM) however you can **notice** the seams on the shadows. Again, you can use techniques but you cannot get SEAMLESS shadows. This idea would make shadows complete. The whole purpose of CSM is to use a depth test, period. Otherwise it's just Angle(0,EyeAngles().y), 0)......

Sources:
https://github.com/Facepunch/garrysmod-requests/issues/765

Similar thread:
https://github.com/Facepunch/garrysmod-requests/issues/1258

**Comment by CaskoChan** [comment]
Is similar to how valve implemented csm in csgo?

**Comment by xoticdev** [comment]
Yes, CSM is used in modern games, in general it's used for a whole bunch of games. CSGO uses CSM most likely. Name any game, and this probably is included without people knowing. Source Engine has baked lighting, and cascades for props from light sources. It's an easy way to cover dynamic looking shadows since the maps never lapse time.

CSM just covers the entire screen with light sources (Ortho matrices) and makes the game appear to have dynamic shadows. It's sort of included anywhere you can look in modern games. Metro Life, Fallout, any game with large landscapes.

**Comment by xoticdev** [comment]
https://github.com/Facepunch/garrysmod-requests/issues/1258

This feature is in request for 2 years, oh wow. Any chance we can get it included soon? :)

**Comment by xoticdev** [comment]
![image](https://cdn.discordapp.com/attachments/402309972333035521/718803088063856670/unknown.png)

As seen here, using mat_showmiplevels 1. this clearly demonstrates depth passes in which I'm referring to so that we can slice up the flashlights and make seamless shadows. This would resolve major shadow issues for maps! Adding a depth filter would be amazing, I must reiterate.

**Comment by xoticdev** [comment]
![image](https://user-images.githubusercontent.com/63826034/97766908-1e224380-1aef-11eb-8b42-9f621f0aec21.png)

Great! New update has a lot of features to check out. I'll follow up on the program and see if I can get better looking results. I'll follow back up on this.

**Comment by xoticdev** [comment]
r_shadows_gamecontrol has been a great help. I'd try to use it to replace the old shadow system but...
Problem with that is that it's a client cheat.
![image](https://user-images.githubusercontent.com/63826034/99011753-cd611080-251a-11eb-856f-e3ec528d9816.png)

**Comment by Xenthio** [comment]
bumping, would like to see this added.
